<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quantitative proteomics data analysis with QFeatures • QFeaturesWorkshop2020</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Quantitative proteomics data analysis with QFeatures">
<meta property="og:description" content="QFeaturesWorkshop2020">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QFeaturesWorkshop2020</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v01-QFeaturesClass.html">The QFeatures class</a>
    </li>
    <li>
      <a href="../articles/v02-QFeaturesPipeline.html">Quantitative proteomics data analysis with QFeatures</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lgatto/QFeaturesWorkshop2020">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="v02-QFeaturesPipeline_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="v02-QFeaturesPipeline_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="v02-QFeaturesPipeline_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quantitative proteomics data analysis with QFeatures</h1>
                        <h4 class="author">Laurent Gatto and Christophe Vanderaa</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/lgatto/QFeaturesWorkshop2020/blob/master/vignettes/v02-QFeaturesPipeline.Rmd"><code>vignettes/v02-QFeaturesPipeline.Rmd</code></a></small>
      <div class="hidden name"><code>v02-QFeaturesPipeline.Rmd</code></div>

    </div>

    
    
<p><strong>Last modified:</strong> NA<br><strong>Compiled</strong>: Mon Dec 14 15:25:26 2020</p>
<p><strong>Learning Objectives</strong> The goals of this workshop are to provide a real-life example of step-by-step quantitative proteomics data analysis using the <code>QFeatures</code> package. This vignette demonstrates an application using spiked-in data.</p>
<div id="analysis-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#analysis-pipeline" class="anchor"></a>Analysis pipeline</h2>
<p>Quantitative proteomics data processing is composed of the following steps:</p>
<ul>
<li>Data import</li>
<li>Exploratory data analysis (PCA)</li>
<li>Missing data management (filtering and/or imputation)</li>
<li>Data cleaning</li>
<li>Transformation and normalisation</li>
<li>Aggregation</li>
<li>Downstream analysis</li>
</ul>
<p>Let’s start by loading the requried packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://tidyverse.tidyverse.org">"tidyverse"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://ggplot2.tidyverse.org">"ggplot2"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/RforMassSpectrometry/QFeatures">"QFeatures"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://bioinf.wehi.edu.au/limma">"limma"</a></span><span class="op">)</span></code></pre></div>
<div id="the-cptac-data" class="section level3">
<h3 class="hasAnchor">
<a href="#the-cptac-data" class="anchor"></a>The CPTAC data</h3>
<p>The CPTAC spike-in study 6 <span class="citation">(Paulovich et al. 2010)</span> combines the Sigma UPS1 standard containing 48 different human proteins that are spiked in at 5 different concentrations (conditions A to E) into a constant yeast protein background. The sample were acquired in triplicate on different instruments in different labs. We are going to start with a subset of the CPTAC study 6 containing conditions A and B for a single lab.</p>
<div class="figure" style="text-align: center">
<img src="cptac.png" alt="The CPTAC spike-in study design (credit Lieven Clement, statOmics, Ghent University)." width="70%"><p class="caption">
The CPTAC spike-in study design (credit Lieven Clement, statOmics, Ghent University).
</p>
</div>
<p>The peptide-level data, as processed by MaxQuant <span class="citation">(Cox and Mann 2008)</span> is available in the <code>msdata</code> package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu">msdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msdata/man/proteomics.html">quant</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"cptac"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co">## [1] "cptac_a_b_peptides.txt"</span></code></pre></div>
<p>From the names of the columns, we see that the quantitative columns, starting with <code>"Intensity."</code> (note the dot!) are at positions 56 to 61.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span><span class="op">)</span>
<span class="co">##  [1] "Sequence"                 "N.term.cleavage.window"  </span>
<span class="co">##  [3] "C.term.cleavage.window"   "Amino.acid.before"       </span>
<span class="co">##  [5] "First.amino.acid"         "Second.amino.acid"       </span>
<span class="co">##  [7] "Second.last.amino.acid"   "Last.amino.acid"         </span>
<span class="co">##  [9] "Amino.acid.after"         "A.Count"                 </span>
<span class="co">## [11] "R.Count"                  "N.Count"                 </span>
<span class="co">## [13] "D.Count"                  "C.Count"                 </span>
<span class="co">## [15] "Q.Count"                  "E.Count"                 </span>
<span class="co">## [17] "G.Count"                  "H.Count"                 </span>
<span class="co">## [19] "I.Count"                  "L.Count"                 </span>
<span class="co">## [21] "K.Count"                  "M.Count"                 </span>
<span class="co">## [23] "F.Count"                  "P.Count"                 </span>
<span class="co">## [25] "S.Count"                  "T.Count"                 </span>
<span class="co">## [27] "W.Count"                  "Y.Count"                 </span>
<span class="co">## [29] "V.Count"                  "U.Count"                 </span>
<span class="co">## [31] "Length"                   "Missed.cleavages"        </span>
<span class="co">## [33] "Mass"                     "Proteins"                </span>
<span class="co">## [35] "Leading.razor.protein"    "Start.position"          </span>
<span class="co">## [37] "End.position"             "Unique..Groups."         </span>
<span class="co">## [39] "Unique..Proteins."        "Charges"                 </span>
<span class="co">## [41] "PEP"                      "Score"                   </span>
<span class="co">## [43] "Identification.type.6A_7" "Identification.type.6A_8"</span>
<span class="co">## [45] "Identification.type.6A_9" "Identification.type.6B_7"</span>
<span class="co">## [47] "Identification.type.6B_8" "Identification.type.6B_9"</span>
<span class="co">## [49] "Experiment.6A_7"          "Experiment.6A_8"         </span>
<span class="co">## [51] "Experiment.6A_9"          "Experiment.6B_7"         </span>
<span class="co">## [53] "Experiment.6B_8"          "Experiment.6B_9"         </span>
<span class="co">## [55] "Intensity"                "Intensity.6A_7"          </span>
<span class="co">## [57] "Intensity.6A_8"           "Intensity.6A_9"          </span>
<span class="co">## [59] "Intensity.6B_7"           "Intensity.6B_8"          </span>
<span class="co">## [61] "Intensity.6B_9"           "Reverse"                 </span>
<span class="co">## [63] "Potential.contaminant"    "id"                      </span>
<span class="co">## [65] "Protein.group.IDs"        "Mod..peptide.IDs"        </span>
<span class="co">## [67] "Evidence.IDs"             "MS.MS.IDs"               </span>
<span class="co">## [69] "Best.MS.MS"               "Oxidation..M..site.IDs"  </span>
<span class="co">## [71] "MS.MS.Count"</span>
<span class="op">(</span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"Intensity\\."</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">## [1] 56 57 58 59 60 61</span></code></pre></div>
<p>We now read these data using the <code>readSummarizedExperiment</code> function. This peptide-level expression data will be imported into R as an instance of class <code>SummarizedExperiment</code>. We also use the <code>fnames</code> argument to set the row-names of the <code>peptides</code> assay to the peptide sequences and specify that the file is a tab-separated table.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cptac_se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html">readSummarizedExperiment</a></span><span class="op">(</span><span class="va">f</span>, ecol <span class="op">=</span> <span class="va">i</span>, fnames <span class="op">=</span> <span class="st">"Sequence"</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="va">cptac_se</span>
<span class="co">## class: SummarizedExperiment </span>
<span class="co">## dim: 11466 6 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(1): ''</span>
<span class="co">## rownames(11466): AAAAGAGGAGDSGDAVTK AAAALAGGK ... YYTVFDRDNNR</span>
<span class="co">##   YYTVFDRDNNRVGFAEAAR</span>
<span class="co">## rowData names(65): Sequence N.term.cleavage.window ...</span>
<span class="co">##   Oxidation..M..site.IDs MS.MS.Count</span>
<span class="co">## colnames(6): Intensity.6A_7 Intensity.6A_8 ... Intensity.6B_8</span>
<span class="co">##   Intensity.6B_9</span>
<span class="co">## colData names(0):</span></code></pre></div>
<p>Before proceeding, we are going to clean up the sample names and annotate the experiment:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"I.+\\."</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">)</span>
<span class="va">cptac_se</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"_[7-9]"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">)</span>
<span class="va">cptac_se</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"^.+_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">)</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span>
<span class="co">## DataFrame with 6 rows and 2 columns</span>
<span class="co">##        condition          id</span>
<span class="co">##      &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 6A_7          6A           7</span>
<span class="co">## 6A_8          6A           8</span>
<span class="co">## 6A_9          6A           9</span>
<span class="co">## 6B_7          6B           7</span>
<span class="co">## 6B_8          6B           8</span>
<span class="co">## 6B_9          6B           9</span></code></pre></div>
<p>Let’s also keep only a subset of</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keep_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sequence"</span>, <span class="st">"Proteins"</span>, <span class="st">"Leading.razor.protein"</span>, <span class="st">"PEP"</span>,
              <span class="st">"Score"</span>, <span class="st">"Reverse"</span>, <span class="st">"Potential.contaminant"</span><span class="op">)</span>

<span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">[</span>, <span class="va">keep_var</span><span class="op">]</span></code></pre></div>
</div>
<div id="missing-values" class="section level3">
<h3 class="hasAnchor">
<a href="#missing-values" class="anchor"></a>Missing values</h3>
<p>Missing values can be highly frequent in proteomics. These exist two reasons supporting the existence of missing values, namely biological or technical.</p>
<ol style="list-style-type: decimal">
<li><p>Values that are missing due to the absence (or extremely low contentration) of a protein are observed for biological reasons, and their pattern <strong>aren’t random</strong>. A protein missing in due to the suppression of its expression will not be missing at random: it will be missing in the condition in which it was suppressed, and be present in the condition where it is expressed.</p></li>
<li><p>Due to it’s data-dependent acquisition, mass spectrometry isn’t capable to assaying all peptides in a sample. Peptides that are less abundant than some of their co-eluting ions, peptides that do not ionise well or peptides that do not get identified might be sporadically missing in the final quantitation table, despite their presence in the biological samples. Their absence patterns are <strong>random</strong> in such cases.</p></li>
</ol>
<p>Often, third party software that produce quantiative data use zeros instead of properly reporting missing values. We can use the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">zeroIsNA()</a></code> function to replace the <code>0</code> by <code>NA</code> values in our <code>cptac_se</code> object and then explore the missing data patterns across columns and rows.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cptac_se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">zeroIsNA</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">nNA</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span>
<span class="co">## $nNA</span>
<span class="co">## DataFrame with 1 row and 2 columns</span>
<span class="co">##         nNA       pNA</span>
<span class="co">##   &lt;integer&gt; &lt;numeric&gt;</span>
<span class="co">## 1     31130   45.2497</span>
<span class="co">## </span>
<span class="co">## $nNArows</span>
<span class="co">## DataFrame with 11466 rows and 3 columns</span>
<span class="co">##                name       nNA       pNA</span>
<span class="co">##         &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;</span>
<span class="co">## 1     AAAAGAGGAG...         4   66.6667</span>
<span class="co">## 2         AAAALAGGK         0    0.0000</span>
<span class="co">## 3        AAAALAGGKK         0    0.0000</span>
<span class="co">## 4     AAADALSDLE...         0    0.0000</span>
<span class="co">## 5     AAADALSDLE...         0    0.0000</span>
<span class="co">## ...             ...       ...       ...</span>
<span class="co">## 11462 YYSIYDLGNN...         6  100.0000</span>
<span class="co">## 11463 YYTFNGPNYN...         3   50.0000</span>
<span class="co">## 11464    YYTITEVATR         4   66.6667</span>
<span class="co">## 11465 YYTVFDRDNN...         6  100.0000</span>
<span class="co">## 11466 YYTVFDRDNN...         6  100.0000</span>
<span class="co">## </span>
<span class="co">## $nNAcols</span>
<span class="co">## DataFrame with 6 rows and 3 columns</span>
<span class="co">##          name       nNA       pNA</span>
<span class="co">##   &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;</span>
<span class="co">## 1        6A_7      4743   41.3658</span>
<span class="co">## 2        6A_8      5483   47.8196</span>
<span class="co">## 3        6A_9      5320   46.3980</span>
<span class="co">## 4        6B_7      4721   41.1739</span>
<span class="co">## 5        6B_8      5563   48.5174</span>
<span class="co">## 6        6B_9      5300   46.2236</span></code></pre></div>
<div class="figure">
<img src="v02-QFeaturesPipeline_files/figure-html/imagena-1.png" alt="Distribution of missing value (white). Peptides row with more missing values are moved towards the top of the figure." width="700"><p class="caption">
Distribution of missing value (white). Peptides row with more missing values are moved towards the top of the figure.
</p>
</div>
<p>Let’s now explore these missing values:</p>
<ul>
<li>Explore the number or proportion of missing values across peptides and samples of the <code>cptac_se</code> data.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">nNA</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">$</span><span class="va">nNAcols</span><span class="op">$</span><span class="va">pNA</span><span class="op">)</span></code></pre></div>
<p><img src="v02-QFeaturesPipeline_files/figure-html/na2-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">nNA</a></span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">$</span><span class="va">nNArows</span><span class="op">$</span><span class="va">nNA</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">##    0    1    2    3    4    5    6 </span>
<span class="co">## 4059  990  884  717  934  807 3075</span></code></pre></div>
<ul>
<li>Remove row that have <em>too many</em> missing values. You can do this by hand or using the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">filterNA()</a></code> function.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## remove rows that have 4 or more NAs out of 6</span>
<span class="va">cptac_se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">filterNA</a></span><span class="op">(</span><span class="va">cptac_se</span>, pNA <span class="op">=</span> <span class="fl">4</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span></code></pre></div>
</div>
<div id="imputation" class="section level3">
<h3 class="hasAnchor">
<a href="#imputation" class="anchor"></a>Imputation</h3>
<p>Imputation is the technique of replacing missing data with probable values. This can be done with <code><a href="https://rdrr.io/pkg/QFeatures/man/impute.html">impute()</a></code> method. As we have discussed above, there are however two types of missing values in mass spectrometry-based proteomics, namely data missing at random (MAR), and data missing not at random (MNAR). These two types of missing data need to be imputed with <a href="https://rformassspectrometry.github.io/QFeatures/articles/Processing.html#imputation-1">different types of imputation methods</a> <span class="citation">(Lazar et al. 2016)</span>.</p>
<div class="figure">
<img src="v02-QFeaturesPipeline_files/figure-html/miximp-1.png" alt="Mixed imputation method. Black cells represent presence of quantitation values and light grey corresponds to missing data. The two groups of interest are depicted in green and blue along the heatmap columns. Two classes of proteins are annotated on the left: yellow are proteins with randomly occurring missing values (if any) while proteins in brown are candidates for non-random missing value imputation." width="700"><p class="caption">
Mixed imputation method. Black cells represent presence of quantitation values and light grey corresponds to missing data. The two groups of interest are depicted in green and blue along the heatmap columns. Two classes of proteins are annotated on the left: yellow are proteins with randomly occurring missing values (if any) while proteins in brown are candidates for non-random missing value imputation.
</p>
</div>
<p>When downstream analyses permit, it might thus be safer not to impute data and deal explicitly with missing values. This is possible when performing hypethesis tests, but not to perform a principal component analysis.</p>
</div>
<div id="identification-quality-control" class="section level3">
<h3 class="hasAnchor">
<a href="#identification-quality-control" class="anchor"></a>Identification quality control</h3>
<p>As discussed in the previous chapter, PSMs are deemed relevant after comparison against hist from a decoy database. The origin of these hits is recorded with <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> in the <code>Reverse</code> variable:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">$</span><span class="va">Reverse</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">##         + </span>
<span class="co">## 7572   12</span></code></pre></div>
<p>Similarly, a proteomics experiment is also searched against a database of contaminants:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span><span class="op">$</span><span class="va">Potential.contaminant</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">##         + </span>
<span class="co">## 7558   26</span></code></pre></div>
<p>Let’s visualise some of the cptac’s metadata using standard <code>ggplot2</code> code:</p>
<ul>
<li>Visualise the score distributions from forward and reverse hits and interpret the figure.</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Score</span>, colour <span class="op">=</span> <span class="va">Reverse</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="v02-QFeaturesPipeline_files/figure-html/idqc1-1.png" width="700"></p>
<ul>
<li>Do the same with the posterior probability score (PEP).</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PEP</span>, colour <span class="op">=</span> <span class="va">Reverse</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="v02-QFeaturesPipeline_files/figure-html/idqc2-1.png" width="700"></p>
</div>
<div id="creating-the-qfeatures-data" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-the-qfeatures-data" class="anchor"></a>Creating the QFeatures data</h3>
<p>We can now create our <code>QFeatures</code> object using the <code>SummarizedExperiment</code> as show below.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cptac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">QFeatures</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>peptides <span class="op">=</span> <span class="va">cptac_se</span><span class="op">)</span><span class="op">)</span>
<span class="va">cptac</span>
<span class="co">## An instance of class QFeatures containing 1 assays:</span>
<span class="co">##  [1] peptides: SummarizedExperiment with 7584 rows and 6 columns</span></code></pre></div>
<p>We should also assign the <code>QFeatures</code> column data with the <code>SummarizedExperiment</code> slot.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">cptac</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cptac_se</span><span class="op">)</span></code></pre></div>
<p>Note that it is also possible to directly create a <code>QFeatures</code> object with the <code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html">readQFeatures()</a></code> function and the same arguments as the <code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html">readSummarizedExperiment()</a></code> used above. In addition, most functions used above and below work on single <code>SummarizedExperiment</code> objects or assays within a <code>QFeatures</code> object.</p>
</div>
<div id="filtering-out-contaminants-and-reverse-hits" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-out-contaminants-and-reverse-hits" class="anchor"></a>Filtering out contaminants and reverse hits</h3>
<p>Using the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures()</a></code> function, filter out the reverse and contaminant hits.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cptac</span> <span class="op">&lt;-</span>
    <span class="va">cptac</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Potential.contaminant</span> <span class="op">!=</span> <span class="st">"+"</span><span class="op">)</span></code></pre></div>
</div>
<div id="log-transformation-and-normaliation" class="section level3">
<h3 class="hasAnchor">
<a href="#log-transformation-and-normaliation" class="anchor"></a>Log-transformation and normaliation</h3>
<p>The two code chunks below log-transform and normalise using the assay <code>i</code> as input and adding a new one names as defined by <code>name</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cptac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">logTransform</a></span><span class="op">(</span><span class="va">cptac</span>, i <span class="op">=</span> <span class="st">"peptides"</span>,
                      name <span class="op">=</span> <span class="st">"log_peptides"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cptac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">normalize</a></span><span class="op">(</span><span class="va">cptac</span>, i <span class="op">=</span> <span class="st">"log_peptides"</span>,
                   name <span class="op">=</span> <span class="st">"lognorm_peptides"</span>,
                   method <span class="op">=</span> <span class="st">"center.median"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cptac</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cptac</span><span class="op">[[</span><span class="st">"log_peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cptac</span><span class="op">[[</span><span class="st">"lognorm_peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v02-QFeaturesPipeline_files/figure-html/plotdens-1.png" alt="Three peptide level assays: raw data, log transformed and normalised." width="1440"><p class="caption">
Three peptide level assays: raw data, log transformed and normalised.
</p>
</div>
</div>
<div id="aggregation" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation" class="anchor"></a>Aggregation</h3>
<p>Below, we are going to use median aggregation, as a first attempt. This is however not the best choice, as we will see later.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cptac</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-aggregate.html">aggregateFeatures</a></span><span class="op">(</span><span class="va">cptac</span>,
                      <span class="st">"lognorm_peptides"</span>,
                      name <span class="op">=</span> <span class="st">"proteins_med"</span>,
                      fcol <span class="op">=</span> <span class="st">"Leading.razor.protein"</span>,
                      fun <span class="op">=</span> <span class="va">colMedians</span>,
                      na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Looking at the <code>.n</code> row variable computed during the aggregation, we see that most proteins result of the aggregation of 5 peptides or less, while very few proteins are accounted for by tens of peptides.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">cptac</span><span class="op">[[</span><span class="st">"proteins_med"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">.n</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">##   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 </span>
<span class="co">## 328 234 162 138  85  72  66  49  46  34  25  28  21  12  17   9   5   4  13   5 </span>
<span class="co">##  21  22  23  24  25  26  28  29  30  31  32  35  38  39  42  43  51  52  62 </span>
<span class="co">##   7   6   4   4   3   3   1   3   1   1   3   1   2   1   1   1   1   1   1</span></code></pre></div>
</div>
<div id="principal-component-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#principal-component-analysis" class="anchor"></a>Principal component analysis</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://www.sthda.com/english/rpkgs/factoextra">"factoextra"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com">"patchwork"</a></span><span class="op">)</span>

<span class="va">pca_pep</span> <span class="op">&lt;-</span>
    <span class="va">cptac</span><span class="op">[[</span><span class="st">"lognorm_peptides"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">filterNA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">assay</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span>scale <span class="op">=</span> <span class="cn">TRUE</span>, center <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html">fviz_pca_ind</a></span><span class="op">(</span>habillage <span class="op">=</span> <span class="va">cptac</span><span class="op">$</span><span class="va">condition</span>, title <span class="op">=</span> <span class="st">"Peptides"</span><span class="op">)</span>

<span class="va">pca_prot</span> <span class="op">&lt;-</span>
    <span class="va">cptac</span><span class="op">[[</span><span class="st">"proteins_med"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">filterNA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">assay</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html">fviz_pca_ind</a></span><span class="op">(</span>habillage <span class="op">=</span> <span class="va">cptac</span><span class="op">$</span><span class="va">condition</span>,
                 title <span class="op">=</span> <span class="st">"Proteins (median aggregation)"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pca_pep</span> <span class="op">+</span> <span class="va">pca_prot</span></code></pre></div>
<div class="figure">
<img src="v02-QFeaturesPipeline_files/figure-html/plotpca-1.png" alt="Peptide and protein level PCA analyses." width="1152"><p class="caption">
Peptide and protein level PCA analyses.
</p>
</div>
</div>
<div id="visualisation" class="section level3">
<h3 class="hasAnchor">
<a href="#visualisation" class="anchor"></a>Visualisation</h3>
<p>Below, we use the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat()</a></code> function to extract the quantitative and row data in a long format, that can be directly reused by the tidyverse tools.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat</a></span><span class="op">(</span><span class="va">cptac</span><span class="op">[</span><span class="st">"P02787ups|TRFE_HUMAN_UPS"</span>, ,
                 <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lognorm_peptides"</span>, <span class="st">"proteins_med"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="va">colname</span><span class="op">)</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">colname</span>, y <span class="op">=</span> <span class="va">value</span>, colour <span class="op">=</span> <span class="va">rowname</span>, shape <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">rowname</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="op">~</span> <span class="va">assay</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"P02787ups|TRFE_HUMAN_UPS"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v02-QFeaturesPipeline_files/figure-html/vis-1.png" alt="Peptide and protein expression profile." width="1152"><p class="caption">
Peptide and protein expression profile.
</p>
</div>
</div>
<div id="statistical-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#statistical-analysis" class="anchor"></a>Statistical analysis</h3>
<p>Below, we are going to perform our statistical analysis on the protein data.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prots</span> <span class="op">&lt;-</span> <span class="va">cptac</span><span class="op">[[</span><span class="st">"proteins_med"</span><span class="op">]</span><span class="op">]</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">prots</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cptac</span><span class="op">)</span></code></pre></div>
<p>The <em><a href="https://bioconductor.org/packages/3.13/limma">limma</a></em> package is the precursor package that enables the consistent application of linear models to normalliy distributed omics data in general, and microarrays in particuar.</p>
<p>The <code>limma</code> package implements an empirical Bayes method that provides borrows information across features to estimate the standard error and calculate (so called moderate) t statistics. This approach is demonstrably more powerful that a standard t-tests when the number of samples is lot.</p>
<p>The code chunk below illstrated how to set up the model, fit it, and apply the empirical Bayes moderation.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://bioinf.wehi.edu.au/limma">"limma"</a></span><span class="op">)</span>
<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">prots</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">prots</span><span class="op">)</span>, <span class="va">design</span><span class="op">)</span>
<span class="co">## Warning: Partial NA coefficients for 23 probe(s)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></code></pre></div>
<p>Finally, the <code><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable()</a></code> function is used the extract the results for the coefficient of interest.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"prots$condition6B"</span>, number <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"protein"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>TP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"ups"</span>, <span class="va">protein</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Note the warning about partial <code>NA</code> coefficients for 23 probes:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">na_coefs</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">pull</span><span class="op">(</span><span class="va">protein</span><span class="op">)</span>
<span class="fu">assay</span><span class="op">(</span><span class="va">prots</span><span class="op">[</span><span class="va">na_coefs</span>, <span class="op">]</span><span class="op">)</span>
<span class="co">##                                6A_7      6A_8       6A_9       6B_7       6B_8</span>
<span class="co">## P00167ups|CYB5_HUMAN_UPS        NaN       NaN        NaN -0.7642817 -2.0050174</span>
<span class="co">## P01112ups|RASH_HUMAN_UPS        NaN       NaN        NaN -1.5367155        NaN</span>
<span class="co">## P05413ups|FABPH_HUMAN_UPS       NaN       NaN        NaN -3.3221739        NaN</span>
<span class="co">## P08758ups|ANXA5_HUMAN_UPS       NaN       NaN        NaN -2.7776131 -1.9904771</span>
<span class="co">## sp|P06704|CDC31_YEAST           NaN       NaN        NaN -1.1834305 -2.1019557</span>
<span class="co">## sp|P32608|RTG2_YEAST            NaN       NaN        NaN        NaN -4.4191375</span>
<span class="co">## sp|P32769|HBS1_YEAST            NaN -1.359851 -0.7138411        NaN        NaN</span>
<span class="co">## sp|P34217|PIN4_YEAST            NaN       NaN        NaN -0.8180873 -0.1083583</span>
<span class="co">## sp|P34237|CASP_YEAST            NaN       NaN        NaN -1.5447431 -1.6367477</span>
<span class="co">## sp|P38166|SFT2_YEAST      -1.564052 -1.052527        NaN        NaN        NaN</span>
<span class="co">## sp|P40056|GET2_YEAST            NaN -1.067516 -1.3866843        NaN        NaN</span>
<span class="co">## sp|P40533|TED1_YEAST            NaN       NaN        NaN -2.0294135        NaN</span>
<span class="co">## sp|P46965|SPC1_YEAST            NaN -3.404591 -3.6174615        NaN        NaN</span>
<span class="co">## sp|P48363|PFD3_YEAST            NaN       NaN        NaN -0.1707164        NaN</span>
<span class="co">## sp|P53044|UFD1_YEAST            NaN       NaN        NaN -2.3740155 -1.4905932</span>
<span class="co">## sp|P53091|MCM6_YEAST            NaN -1.229600 -1.6762933        NaN        NaN</span>
<span class="co">## sp|P53115|INO80_YEAST           NaN       NaN        NaN -2.6986219        NaN</span>
<span class="co">## sp|P53893|RIA1_YEAST      -1.809222       NaN -1.5085444        NaN        NaN</span>
<span class="co">## sp|Q05029|BCH1_YEAST      -1.836222 -1.736322        NaN        NaN        NaN</span>
<span class="co">## sp|Q07540|FRDA_YEAST            NaN       NaN        NaN -2.3652501        NaN</span>
<span class="co">## sp|Q08645|FOLE_YEAST            NaN       NaN        NaN -1.7071443        NaN</span>
<span class="co">## sp|Q12255|NYV1_YEAST            NaN       NaN        NaN -2.8508289        NaN</span>
<span class="co">## sp|Q12280|IQG1_YEAST            NaN       NaN        NaN -2.5438648 -2.0314751</span>
<span class="co">##                                 6B_9</span>
<span class="co">## P00167ups|CYB5_HUMAN_UPS  -1.1067954</span>
<span class="co">## P01112ups|RASH_HUMAN_UPS  -1.5455337</span>
<span class="co">## P05413ups|FABPH_HUMAN_UPS -3.8744226</span>
<span class="co">## P08758ups|ANXA5_HUMAN_UPS -2.0731897</span>
<span class="co">## sp|P06704|CDC31_YEAST     -1.5681249</span>
<span class="co">## sp|P32608|RTG2_YEAST      -2.7710331</span>
<span class="co">## sp|P32769|HBS1_YEAST             NaN</span>
<span class="co">## sp|P34217|PIN4_YEAST      -0.1826536</span>
<span class="co">## sp|P34237|CASP_YEAST      -1.6714608</span>
<span class="co">## sp|P38166|SFT2_YEAST             NaN</span>
<span class="co">## sp|P40056|GET2_YEAST             NaN</span>
<span class="co">## sp|P40533|TED1_YEAST      -1.7311957</span>
<span class="co">## sp|P46965|SPC1_YEAST             NaN</span>
<span class="co">## sp|P48363|PFD3_YEAST      -0.4924892</span>
<span class="co">## sp|P53044|UFD1_YEAST             NaN</span>
<span class="co">## sp|P53091|MCM6_YEAST             NaN</span>
<span class="co">## sp|P53115|INO80_YEAST     -2.7341224</span>
<span class="co">## sp|P53893|RIA1_YEAST             NaN</span>
<span class="co">## sp|Q05029|BCH1_YEAST             NaN</span>
<span class="co">## sp|Q07540|FRDA_YEAST      -2.0162325</span>
<span class="co">## sp|Q08645|FOLE_YEAST      -1.5094503</span>
<span class="co">## sp|Q12255|NYV1_YEAST      -2.8074462</span>
<span class="co">## sp|Q12280|IQG1_YEAST      -2.4485187</span></code></pre></div>
<p>We can now visualise the results using a volcano plot:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">adj.P.Val</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">TP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Warning: Removed 23 rows containing missing values (geom_point).</span></code></pre></div>
<div class="figure">
<img src="v02-QFeaturesPipeline_files/figure-html/vp-1.png" alt="Volcano plot highlighing spiked-in proteins in red." width="700"><p class="caption">
Volcano plot highlighing spiked-in proteins in red.
</p>
</div>
<p>Using the pipeline described above, we would would identify a single differentially expressed protein at an 5 percent FDR but miss out the other 36 expected spike-in proteins.</p>
<p>We can assess our results in terms of true/false postitves/negatives:</p>
<ul>
<li>True positives: 1</li>
<li>False positives: 0</li>
<li>True negatives: 1342</li>
<li>False negatives: 32</li>
</ul>
</div>
</div>
<div id="final-exercise" class="section level2">
<h2 class="hasAnchor">
<a href="#final-exercise" class="anchor"></a>Final exercise</h2>
<p>As shown below, it is possible to substantially improve these results by aggregating features using a robust summarisation (available as <code><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html">MsCoreUtils::robustSummary()</a></code>), i.e robust regression with M-estimation using Huber weights, as described in section 2.7 in <span class="citation">(Sticker et al. 2019)</span>.</p>
<div class="figure">
<img src="vp2.png" alt=""><p class="caption">Aggregation using robust summarisation.</p>
</div>
<ul>
<li>True positives: 21</li>
<li>False positives: 2</li>
<li>True negatives: 1340</li>
<li>False negatives: 12</li>
</ul>
<p>Repeat and adapt what we have seen here using, for example, the <code>robustSummary()</code> function.</p>
</div>
<div id="session-information" class="section level2">
<h2 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h2>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## R Under development (unstable) (2020-11-02 r79396)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.1 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] patchwork_1.1.0             factoextra_1.0.7           </span>
<span class="co">##  [3] gplots_3.1.1                limma_3.47.1               </span>
<span class="co">##  [5] QFeatures_1.1.0             MultiAssayExperiment_1.17.2</span>
<span class="co">##  [7] SummarizedExperiment_1.21.1 Biobase_2.51.0             </span>
<span class="co">##  [9] GenomicRanges_1.43.1        GenomeInfoDb_1.27.3        </span>
<span class="co">## [11] IRanges_2.25.5              S4Vectors_0.29.6           </span>
<span class="co">## [13] BiocGenerics_0.37.0         MatrixGenerics_1.3.0       </span>
<span class="co">## [15] matrixStats_0.57.0          forcats_0.5.0              </span>
<span class="co">## [17] stringr_1.4.0               dplyr_1.0.2                </span>
<span class="co">## [19] purrr_0.3.4                 readr_1.4.0                </span>
<span class="co">## [21] tidyr_1.1.2                 tibble_3.0.4               </span>
<span class="co">## [23] ggplot2_3.3.2               tidyverse_1.3.0            </span>
<span class="co">## [25] BiocStyle_2.19.1           </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##  [1] colorspace_2.0-0        ggsignif_0.6.0          ellipsis_0.3.1         </span>
<span class="co">##  [4] rio_0.5.16              rprojroot_2.0.2         XVector_0.31.1         </span>
<span class="co">##  [7] fs_1.5.0                rstudioapi_0.13         ggpubr_0.4.0           </span>
<span class="co">## [10] farver_2.0.3            ggrepel_0.8.2           fansi_0.4.1            </span>
<span class="co">## [13] lubridate_1.7.9.2       xml2_1.3.2              knitr_1.30             </span>
<span class="co">## [16] jsonlite_1.7.2          broom_0.7.2             dbplyr_2.0.0           </span>
<span class="co">## [19] BiocManager_1.30.10     compiler_4.1.0          httr_1.4.2             </span>
<span class="co">## [22] backports_1.2.1         assertthat_0.2.1        Matrix_1.2-18          </span>
<span class="co">## [25] lazyeval_0.2.2          cli_2.2.0               htmltools_0.5.0        </span>
<span class="co">## [28] tools_4.1.0             gtable_0.3.0            glue_1.4.2             </span>
<span class="co">## [31] GenomeInfoDbData_1.2.4  Rcpp_1.0.5              carData_3.0-4          </span>
<span class="co">## [34] cellranger_1.1.0        pkgdown_1.6.1           vctrs_0.3.5            </span>
<span class="co">## [37] xfun_0.19               ps_1.5.0                openxlsx_4.2.3         </span>
<span class="co">## [40] rvest_0.3.6             lifecycle_0.2.0         gtools_3.8.2           </span>
<span class="co">## [43] rstatix_0.6.0           zlibbioc_1.37.0         MASS_7.3-53            </span>
<span class="co">## [46] scales_1.1.1            ragg_0.4.0              hms_0.5.3              </span>
<span class="co">## [49] ProtGenerics_1.23.5     AnnotationFilter_1.15.0 curl_4.3               </span>
<span class="co">## [52] yaml_2.2.1              memoise_1.1.0           stringi_1.5.3          </span>
<span class="co">## [55] highr_0.8               desc_1.2.0              caTools_1.18.0         </span>
<span class="co">## [58] zip_2.1.1               rlang_0.4.9             pkgconfig_2.0.3        </span>
<span class="co">## [61] systemfonts_0.3.2       bitops_1.0-6            evaluate_0.14          </span>
<span class="co">## [64] lattice_0.20-41         labeling_0.4.2          tidyselect_1.1.0       </span>
<span class="co">## [67] magrittr_2.0.1          R6_2.5.0                generics_0.1.0         </span>
<span class="co">## [70] DelayedArray_0.17.5     DBI_1.1.0               foreign_0.8-80         </span>
<span class="co">## [73] pillar_1.4.7            haven_2.3.1             withr_2.3.0            </span>
<span class="co">## [76] MsCoreUtils_1.3.1       abind_1.4-5             RCurl_1.98-1.2         </span>
<span class="co">## [79] msdata_0.31.0           modelr_0.1.8            crayon_1.3.4           </span>
<span class="co">## [82] car_3.0-10              KernSmooth_2.23-18      rmarkdown_2.5          </span>
<span class="co">## [85] grid_4.1.0              readxl_1.3.1            data.table_1.13.4      </span>
<span class="co">## [88] reprex_0.3.0            digest_0.6.27           textshaping_0.2.1      </span>
<span class="co">## [91] munsell_0.5.0</span></code></pre></div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Cox:2008">
<p>Cox, J, and M Mann. 2008. “MaxQuant Enables High Peptide Identification Rates, Individualized P.p.b.-range Mass Accuracies and Proteome-Wide Protein Quantification.” <em>Nat Biotechnol</em> 26 (12): 1367–72. <a href="https://doi.org/10.1038/nbt.1511">https://doi.org/10.1038/nbt.1511</a>.</p>
</div>
<div id="ref-Lazar:2016">
<p>Lazar, C, L Gatto, M Ferro, C Bruley, and T Burger. 2016. “Accounting for the Multiple Natures of Missing Values in Label-Free Quantitative Proteomics Data Sets to Compare Imputation Strategies.” <em>J Proteome Res</em> 15 (4): 1116–25. <a href="https://doi.org/10.1021/acs.jproteome.5b00981">https://doi.org/10.1021/acs.jproteome.5b00981</a>.</p>
</div>
<div id="ref-Paulovich:2010">
<p>Paulovich, Amanda G, Dean Billheimer, Amy-Joan L Ham, Lorenzo Vega-Montoto, Paul A Rudnick, David L Tabb, Pei Wang, et al. 2010. “Interlaboratory Study Characterizing a Yeast Performance Standard for Benchmarking LC-MS Platform Performance.” <em>Mol. Cell. Proteomics</em> 9 (2): 242–54.</p>
</div>
<div id="ref-Sticker:2019">
<p>Sticker, Adriaan, Ludger Goeminne, Lennart Martens, and Lieven Clement. 2019. “Robust Summarization and Inference in Proteome-Wide Label-Free Quantification.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/668863">https://doi.org/10.1101/668863</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
